src/Segwit.mo

```diff
@@ -0,0 +1,152 @@
+import Nat "mo:base/Nat";
+import Nat8 "mo:base/Nat8";
+import Nat32 "mo:base/Nat32";
+import Result "mo:base/Result";
+import Buffer "mo:base/Buffer";
+import Iter "mo:base/Iter";
+import P "mo:base/Prelude";
+import Bech32 "../src/Bech32";
+
+module {
+
+  public type WitnessProgram = {
+    version : Nat8;
+    program : [Nat8];
+  };
+
+  // Convert a Witness Program to a SegWit Address.
+  public func encode(hrp : Text, {version; program} : WitnessProgram) :
+    Result.Result<Text, Text> {
+
+    let bech32Input = Buffer.Buffer<Nat8>(program.size());
+    bech32Input.add(version);
+
+    switch (convertBits(program.vals(), bech32Input, 8, 5, true)) {
+      case (#err (msg)) {
+        return #err (msg);
+      };
+      case _ {};
+    };
+
+    let encoding : Bech32.Encoding = if (version > 0) {
```

Comment:
Does this logic conform to the Bitcoin specification?
If new variants are added later, wouldn't there be a risk that the wrong variant is chosen here?

Reply:
Good question, I just mimicked the reference implementation here without much thought, I'm going to verify.

Reply:
This is defined in BIP350: https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki#addresses-for-segregated-witness-outputs and was added to Bitcoin Core: https://github.com/bitcoin/bitcoin/pull/20861.

Reply:
Looks good. Thanks for checking!